### PR TITLE
alphonse: extend STRING-sep SOTA to 18 epochs (PR #311 was epoch-limited)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

PR #311 (edward, STRING-separable learnable position encoding) is the current SOTA on `yi`:
**val_abupt 7.546% / test_abupt 8.771%** (W&B run `gcwx9yaa`, group `tay-round18-grape-ablation`, Arm B).

Looking at the PR #311 Arm B val curve, **the run was clearly epoch-limited**:

| epoch | val_abupt |
|-------|-----------|
| 6     | 8.478%    |
| 7     | 8.159%    |
| 8     | 7.909%    |
| 9     | 7.742% (terminal) |

All val slopes were still negative at the terminal epoch (Δ ep8→ep9 = −0.163pp), meaning the model
had not converged. The reported best of 7.546% reflects the EMA / best-checkpoint effect, but the
underlying loss surface still had headroom.

**Hypothesis:** running the **identical PR #311 stack for 18 epochs (2× budget)** will push
val_abupt below 7.5% without any architectural or HP change. This is the cheapest possible win:
no new code, no tuning, just more steps on a known-stable convergence trajectory.

If the slope continues at ~−0.15 pp/epoch for several more epochs, we should land in the
~6.8–7.2% range, with proportional improvements on the dominant remaining gaps:
**wall_shear_y/z (currently 2.5×, 2.9× AB-UPT targets)** and **volume_pressure (2.0× target)**.
These late-converging axes typically benefit most from extended training.

## Instructions

You are running the **identical** PR #311 winning configuration but for 18 epochs instead of 9.
Do NOT change any other hyperparameter. The only change vs the SOTA run is `--epochs 18`
and a corresponding `--lr-warmup-steps` adjustment so warmup remains 1 epoch.

### Step 1: pull the latest yi (PR #311 STRING-sep code is required)

```bash
cd target/
git fetch origin yi
git pull origin yi
git log --oneline -5  # verify PR #311 merge commit is present
```

You should see PR #311 (edward, STRING-separable PE) in recent commits. The flag
`--pos-encoding-mode string` and the `StringEncoding` class must exist in `train.py`.
If they don't, stop and ping the advisor.

### Step 2: run the experiment (8-GPU torchrun, 18 epochs)

```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent alphonse \
  --wandb-group alphonse-longer-epochs-r25 \
  --optimizer lion \
  --lr 1e-4 \
  --weight-decay 5e-4 \
  --no-compile-model \
  --batch-size 4 \
  --validation-every 1 \
  --train-surface-points 65536 \
  --eval-surface-points 65536 \
  --train-volume-points 65536 \
  --eval-volume-points 65536 \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 4 \
  --model-slices 128 \
  --ema-decay 0.999 \
  --epochs 18 \
  --lr-warmup-steps 2720 \
  --clip-grad-norm 0.5 \
  --pos-encoding-mode string \
  --rff-num-features 32 \
  --rff-sigma 1.0
```

Notes:
- `--lr-warmup-steps 2720` ≈ 1 epoch of warmup (matches PR #311). Cosine decay covers the full 18 epochs.
- `--clip-grad-norm 0.5` matches PR #309 / #311 stack (Lion stability).
- `--model-heads 4` matches PR #311 winning arm (NOT 8).
- `--pos-encoding-mode string` activates the STRING-separable learnable PE from PR #311.
- Use `--agent alphonse` so checkpoints land in the right namespace.

### Step 3: kill criteria

If `val_primary/abupt_axis_mean_rel_l2_pct` is **above 9.5%** by epoch 8 (i.e. clearly worse than
the PR #311 trajectory), stop the run early — it indicates seed variance / regression and we
should debug rather than burn the full 18 epochs.

If gnorm collapses below 300 sustained, also stop and report.

### Step 4: report results

Mark the PR ready for review with the W&B run ID, group, and the **per-epoch val_abupt
trajectory** (epochs 1–18). Include final test metrics from best-validation checkpoint:
`test_primary/abupt_axis_mean_rel_l2_pct`, `surface_pressure_rel_l2_pct`,
`wall_shear_rel_l2_pct`, `volume_pressure_rel_l2_pct`,
and the three wall_shear axes.

## Baseline (must beat)

Current SOTA (PR #311 edward, STRING-sep PE):

| Metric | val | test |
|---|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | **7.546%** | **8.771%** |
| `surface_pressure_rel_l2_pct` | — | 4.485% |
| `wall_shear_rel_l2_pct` | — | 8.227% |
| `volume_pressure_rel_l2_pct` | — | 12.438% |
| `wall_shear_x_rel_l2_pct` | — | 7.253% |
| `wall_shear_y_rel_l2_pct` | — | 9.233% |
| `wall_shear_z_rel_l2_pct` | — | 10.449% |

W&B run: `gcwx9yaa`, group `tay-round18-grape-ablation`, Arm B.

**Merge bar: val_abupt < 7.546%.** Even a small improvement (e.g. 7.4%) compounds.

## Why this is high-EV

1. **No code change required** — pure training-budget experiment, lowest possible execution risk.
2. **Strong prior**: terminal epoch of PR #311 had Δ = −0.163pp/epoch with no sign of a knee.
3. **Targets the worst remaining gaps**: wall_shear_y/z and volume_pressure are typically
   late-converging axes; extra epochs disproportionately help them.
4. **Diagnostic value even on a tie**: if 18 epochs does NOT improve val_abupt, that's strong
   evidence the STRING-sep architecture has a hard floor near 7.5% and we should pivot to
   architecture/loss changes rather than budget extensions.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT alphonse:

Stopping per the kill-switch in Step 1 of the instructions: **PR #311's STRING-sep code is not present on `yi`**.

### Verification

```bash
$ git log origin/yi --oneline | head -5
776afd8 research: update baseline — PR #311 edward STRING-separable PE (val 7.546%, test 8.771%)
a6b60b3 research: assign PR #364 frieren DropPath sweep r24 + close PR #338 (non-responsive)
c0ae62a research: log Round 23 - PR #298 + #297 closures, PR #362 fern + PR #363 haku assignments
...
$ git branch -a --contains 1e5f6fe   # PR #311 merge commit
  remotes/origin/edward/string-sep-extended-epochs
  remotes/origin/fern/surface-tangent-frame-tau
  remotes/origin/frieren/string-sep-volume-decoder
  remotes/origin/nezuko/model-layers5-string-sep
  remotes/origin/tay
  remotes/origin/thorfinn/string-sep-qknorm-stack
```

PR #311 was merged into `tay`, not `yi`. The `776afd8` commit on yi only updates `BASELINE.md` text — it does not include the actual `train.py` code change for `--pos-encoding-mode string` / `StringEncoding`. `yi` also still has the monolithic `train.py` (no `model.py` / `trainer_runtime.py`), which differs structurally from `tay`.

Grep on yi confirms:

```bash
$ grep -E 'pos.encoding.mode|StringEncoding' target/train.py
(no matches)
```

### Comparison: PR #357 (edward, the parallel "extended epochs" PR) targets `tay`, where the code exists. My PR #376 targets `yi`.

### Options for the advisor

1. **Merge `tay` → `yi`** (or cherry-pick PR #311 + dependencies onto yi), then re-fire me.
2. **Re-base PR #376 on `tay`** (changes the advisor branch label).
3. **Close PR #376** in favor of edward's PR #357 if they are duplicates.

Standing by until guidance arrives.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #376 because Closing this PR — the STRING-sep code (PR #311) was merged into the `tay` branch, not `yi`. There is no `--pos-encoding-mode string` flag on `yi/train.py`, so the experiment cannot run as written.

Great diagnostic work, alphonse — you correctly identified the branch mismatch and raised the options clearly. I'm assigning you a fresh experiment on `yi` that is compatible with the current codebase..
